test/functional/test_framework/test_node.py

```diff
@@ -293,8 +295,18 @@ def wait_for_cookie_credentials(self):
         self._raise_assertion_error("Unable to retrieve cookie credentials after {}s".format(self.rpc_timeout))
 
     def generate(self, nblocks, maxtries=1000000):
-        self.log.debug("TestNode.generate() dispatches `generate` call to `generatetoaddress`")
-        return self.generatetoaddress(nblocks=nblocks, address=self.get_deterministic_priv_key().address, maxtries=maxtries)
+        # if wallet is not compiled, blocks are generated to an address or descriptor using an HD seed
+        hd_address = self.getnewaddress()
+        hd_info = self.getaddressinfo(hd_address)
+        if self.getwalletinfo()["descriptors"]:
+            #TODO: figure out how to make sure the keypath does increase & is unique
+            assert_equal(hd_info["hdkeypath"], "m/84'/1'/0'/0/" + str(self.hdkeypathnonce))
+            self.hdkeypathnonce += 1 
```

Comment:
I added `hdkeypathnonce` here to make sure a new key path is used. But I realized that incrementing the counter variable like this won't work, because many test files would call `getnewaddress()` outside of `generate()`. These new keys generated out of band also belong to the family of our HD tree, but our counter variable will not be aware of this. 
(like feature_segwit.py). 

Hence, although we want to make sure each time a new key path is used, it doesn't make sense to use member variable like this. Maybe the best way to check is to compare the new keypath with the existing keypool? Or do you have other suggestions? @jnewbery